src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+#pragma once
+
+#include "consensus/ledger_enclave_types.h"
+#include "ds/logger.h"
+#include "ds/messaging.h"
+
+#include <cstdint>
+#include <cstdio>
+#include <filesystem>
+#include <limits>
+#include <list>
+#include <map>
+#include <string>
+#include <sys/types.h>
+#include <unistd.h>
+#include <vector>
+
+namespace fs = std::filesystem;
+
+namespace asynchost
+{
+  static constexpr size_t max_chunk_threshold_size =
+    std::numeric_limits<uint32_t>::max(); // 4GB
+  static constexpr size_t max_read_cache_size_default = 5;
+  static constexpr auto ledger_committed_suffix = ".committed";
+  static constexpr auto ledger_start_idx_delimiter = "_";
+
+  static inline bool is_ledger_file_compacted(const std::string& file_name)
+  {
+    auto pos = file_name.find(ledger_committed_suffix);
+    return !(pos == std::string::npos);
+  }
+
+  static inline size_t get_start_idx_from_file_name(
+    const std::string& file_name)
+  {
+    auto pos = file_name.find(ledger_start_idx_delimiter);
+    if (pos == std::string::npos)
+    {
+      throw std::logic_error(fmt::format(
+        "Ledger file name {} does not contain a start idx", file_name));
+    }
+
+    return std::stol(file_name.substr(pos + 1));
+  }
+
+  class LedgerFile
+  {
+  private:
+    using positions_offset_header_t = size_t;
+
+    static constexpr auto file_name_prefix = "ledger";
+
+    static constexpr size_t frame_header_size = sizeof(uint32_t);
+
+    const std::string dir;
+
+    size_t start_idx = 1;
+    size_t total_len = 0;
+    std::vector<uint32_t> positions;
+
+    // This uses C stdio instead of fstream because an fstream
+    // cannot be truncated.
+    FILE* file;
+    bool completed = false;
+    bool compacted = false;
+
+  public:
+    LedgerFile(const std::string& dir, size_t start_idx) :
+      dir(dir),
+      file(NULL),
+      start_idx(start_idx)
+    {
+      file = fopen(
+        (fs::path(dir) /
+         fs::path(fmt::format("{}_{}", file_name_prefix, start_idx)))
+          .c_str(),
+        "w+b");
+
+      // Header reserved for the offset to the position table
+      fseeko(file, sizeof(positions_offset_header_t), SEEK_SET);
+      total_len = sizeof(positions_offset_header_t);
+    }
+
+    // Used when recovering an existing ledger file
+    LedgerFile(const std::string& dir, const std::string& file_name) :
+      dir(dir),
+      file(NULL)
+    {
+      auto full_path = (fs::path(dir) / fs::path(file_name));
+      file = fopen(full_path.c_str(), "r+b");
+      if (!file)
+      {
+        throw std::logic_error(
+          fmt::format("Unable to open ledger file {}", full_path));
+      }
+
+      compacted = is_ledger_file_compacted(file_name);
+      start_idx = get_start_idx_from_file_name(file_name);
+
+      // First, get full size of file
+      fseeko(file, 0, SEEK_END);
+      size_t total_file_size = ftello(file);
+
+      if (total_file_size == 0)
+      {
+        // If the file is empty, initialise it as if it were new
```

Comment:
I don't understand this case. If this is during recovery, why is it empty?

Reply:
Good spot! It is indeed an outdated edge case that I forgot to remove.